src/client/pages/EventPage/tabs/TeamsTab.tsx

```diff
@@ -233,6 +243,48 @@ export default class TeamsTab extends EventPageTab<TeamsTabProps, TeamsTabState>
       });
   }
 
+  onTeamEdit = (team: TESCTeam) => {
+    this.setState({
+      editModalState: {
+        isOpen: true,
+        team,
+      }
+    })
+  }
+
+  handleTeamEditSubmit = async (newTeam: TESCTeam) => {
+    const oldTeam = this.state.editModalState.team;
+    const deletedMembersEmails = oldTeam.members.filter(user => !newTeam.members.includes(user)).map(user => user.account!.email!)
+    const addedMembersEmails = newTeam.members.filter(user => !oldTeam.members.includes(user)).map(user => user.account!.email!)
+
+    const removedAllOldMembers = deletedMembersEmails.length === oldTeam.members.length;
+    if (removedAllOldMembers && addedMembersEmails.length > 0) {
+      if (oldTeam.members.length === 4) {
+        // Can't add a new member when the team is full
+        const deletedUser = deletedMembersEmails.pop();
+        await removeUsersFromTeam([deletedUser], this.props.event._id!, oldTeam._id!)
+      }
+
+      // We need to add one member before removing everyone else, so that we don't delete the team due to no members
+      const newUser = addedMembersEmails.pop();
+      await addUsersToTeam([newUser], this.props.event._id!, newTeam._id!)
+    }
+
+    deletedMembersEmails.length > 0 && await removeUsersFromTeam(deletedMembersEmails, this.props.event._id!, oldTeam._id!)
+    addedMembersEmails.length > 0 && await addUsersToTeam(addedMembersEmails, this.props.event._id!, oldTeam._id!)
+    this.onTeamModalToggle();
+    this.props.onTeamsUpdate();
```

Comment:
Tracking updates by breaking them into Adds / Deletes on the frontend feels brittle because you are surfacing up what should be a backend implementation detail into the network - what happens if removeUsersFromTeam times out / fails? Or what happens if removeUsersFromTeam passes but addUserstoTeam fails? It is very easy to create some wonky team state on the backend this way. There is no way to ensure the dependence of HTTP calls without significant overhead - this is not a problem on the backend (https://en.wikipedia.org/wiki/Database_transaction).

I was wondering if you considered naively inserting what the admin tells you to insert instead of trying to "backtrack" their actions (In this case, you are trying to figure out what the admin did on the UI and making the corresponding changes in the system). What if you took a list of 4 emails to the backend, and the backend iterates through them, figures out if they are valid and insert the team if all input was valid? Maybe I'm missing something here - let me know what you think.